File: Examples/test-suite/perl5/overload_template_runme.pl
use strict;
use warnings;
use Test::More tests => 46;
BEGIN { use_ok('overload_template') }
require_ok('overload_template');

my $f = overload_template::foo();

my $a = overload_template::maximum(3, 4);
my $b = overload_template::maximum(3.4, 5.2);

# mix 1
is(overload_template::mix1('hi'), 101, 'mix1(const char *)');
is(overload_template::mix1(1.0, 1.0), 102, 'mix1(double, const double &)');
is(overload_template::mix1(1.0), 103, 'mix1(double)');

# mix 2
is(overload_template::mix2('hi'), 101, 'mix2(const char *)');
is(overload_template::mix2(1.0, 1.0), 102, 'mix2(double, const double &)');
is(overload_template::mix2(1.0), 103, 'mix2(double)');

# mix 3
is(overload_template::mix3('hi'), 101, 'mix3(const char *)');
is(overload_template::mix3(1.0, 1.0), 102, 'mix3(double, const double &)');
is(overload_template::mix3(1.0), 103, 'mix3(double)');

# Combination 1
is(overload_template::overtparams1(100), 10, 'overtparams1(int)');
is(overload_template::overtparams1(100.0, 100), 20, 'overtparams1(double, int)');

# Combination 2
is(overload_template::overtparams2(100.0, 100), 40, 'overtparams2(double, int)');

# Combination 3
is(overload_template::overloaded(), 60, 'overloaded()');
is(overload_template::overloaded(100.0, 100), 70, 'overloaded(double, int)');

# Combination 4
is(overload_template::overloadedagain('hello'), 80, 'overloadedagain(const char *)');
is(overload_template::overloadedagain(), 90, 'overloadedagain(double)');

# specializations
is(overload_template::specialization(10), 202, 'specialization(int)');
is(overload_template::specialization(10.0), 203, 'specialization(double)');
is(overload_template::specialization(10, 10), 204, 'specialization(int, int)');
is(overload_template::specialization(10.0, 10.0), 205, 'specialization(double, double)');
is(overload_template::specialization('hi', 'hi'), 201, 'specialization(const char *, const char *)');

# simple specialization
can_ok('overload_template', 'xyz');
can_ok('overload_template', 'xyz_int');
can_ok('overload_template', 'xyz_double');
overload_template::xyz();
overload_template::xyz_int();
overload_template::xyz_double();

# a bit of everything
is(overload_template::overload('hi'), 0, 'overload()');
is(overload_template::overload(1), 10, 'overload(int t)');
is(overload_template::overload(1, 1), 20, 'overload(int t, const int &)');
is(overload_template::overload(1, 'hello'), 30, 'overload(int t, const char *)');

my $k = overload_template::Klass->new();
is(overload_template::overload($k), 10, 'overload(Klass t)');
is(overload_template::overload($k, $k), 20, 'overload(Klass t, const Klass &)');
is(overload_template::overload($k, 'hello'), 30, 'overload(Klass t, const char *)');
is(overload_template::overload(10.0, 'hi'), 40, 'overload(double t, const char *)');
is(overload_template::overload(), 50, 'overload(const char *)');

# everything put in a namespace
is(overload_template::nsoverload('hi'), 1000, 'nsoverload()');
is(overload_template::nsoverload(1), 1010, 'nsoverload(int t)');
is(overload_template::nsoverload(1, 1), 1020, 'nsoverload(int t, const int &)');
is(overload_template::nsoverload(1, 'hello'), 1030, 'nsoverload(int t, const char *)');
is(overload_template::nsoverload($k), 1010, 'nsoverload(Klass t)');
is(overload_template::nsoverload($k, $k), 1020, 'nsoverload(Klass t, const Klass &)');
is(overload_template::nsoverload($k, 'hello'), 1030, 'nsoverload(Klass t, const char *)');
is(overload_template::nsoverload(10.0, 'hi'), 1040, 'nsoverload(double t, const char *)');
is(overload_template::nsoverload(), 1050, 'nsoverload(const char *)');

can_ok('overload_template::A', 'foo');
overload_template::A->foo(1);
can_ok('overload_template::B', 'foo');
overload_template::B->new()->foo(1);

File: Examples/test-suite/ocaml/overload_template_runme.ml
open Swig
open Overload_template

let _ =
  assert (_foo '() as int = 3);
  assert (_maximum '(3, 4) as int = 4);
  assert (_maximum '(3.4, 5.2) as float > 5.);
  assert (_mix1 '("hi") as int = 101);
  assert (_mix1 '(1.0, 1.0) as int = 102);
  assert (_mix1 '(1.0) as int = 103);
  assert (_mix2 '("hi") as int = 101);
  assert (_mix2 '(1.0, 1.0) as int = 102);
  assert (_mix2 '(1.0) as int = 103);
  assert (_mix3 '("hi") as int = 101);
  assert (_mix3 '(1.0, 1.0) as int = 102);
  assert (_mix3 '(1.0) as int = 103);

  assert (_overtparams1 '(100) as int = 10);
  assert (_overtparams1 '(100.0, 100) as int = 20);
  assert (_overtparams2 '(100.0, 100) as int = 40);
  assert (_overloaded '() as int = 60);
  assert (_overloaded '(100.0, 100) as int = 70);
  assert (_overloadedagain '("hello") as int = 80);
  assert (_overloadedagain '() as int = 90);

  assert (_specialization '(10) as int = 202);
  assert (_specialization '(10.0) as int = 203);
  assert (_specialization '(10, 10) as int = 204);
  assert (_specialization '(10.0, 10.0) as int = 205);
  assert (_specialization '("hi", "hi") as int = 201);

  assert (_xyz '() = C_void);
  assert (_xyz_int '() = C_void);
  assert (_xyz_double '() = C_void);

  assert (_overload '("hi") as int = 0);
  assert (_overload '(1) as int = 10);
  assert (_overload '(1, 1) as int = 20);
  assert (_overload '(1, "hello") as int = 30);
  let k = new_Klass '() in
  assert (_overload '(k) as int = 10);
  assert (_overload '(k, k) as int = 20);
  assert (_overload '(k, "hello") as int = 30);
  assert (_overload '(10.0, "hi") as int = 40);
  assert (_overload '() as int = 50);

  assert (_nsoverload '("hi") as int = 1000);
  assert (_nsoverload '(1) as int = 1010);
  assert (_nsoverload '(1, 1) as int = 1020);
  assert (_nsoverload '(1, "hello") as int = 1030);
  assert (_nsoverload '(k) as int = 1010);
  assert (_nsoverload '(k, k) as int = 1020);
  assert (_nsoverload '(k, "hello") as int = 1030);
  assert (_nsoverload '(10.0, "hi") as int = 1040);
  assert (_nsoverload '() as int = 1050);

  assert (_A_foo '(1) = C_void);
  let b = new_B '() in
  assert (b -> foo(1) = C_void);
;;
